src/main/telemetry/smartport.c

```diff
@@ -145,238 +147,227 @@ const uint16_t frSkyDataIdTable[] = {
 static serialPort_t *smartPortSerialPort = NULL; // The 'SmartPort'(tm) Port.
 static serialPortConfig_t *portConfig;
 
-static bool smartPortTelemetryEnabled =  false;
 static portSharing_e smartPortPortSharing;
 
 char smartPortState = SPSTATE_UNINITIALIZED;
-static uint8_t smartPortHasRequest = 0;
 static uint8_t smartPortIdCnt = 0;
-static uint32_t smartPortLastRequestTime = 0;
 
 typedef struct smartPortFrame_s {
     uint8_t  sensorId;
-    uint8_t  frameId;
-    uint16_t valueId;
-    uint32_t data;
+    smartPortPayload_t payload;
     uint8_t  crc;
 } __attribute__((packed)) smartPortFrame_t;
 
-#define SMARTPORT_FRAME_SIZE  sizeof(smartPortFrame_t)
+#define SMARTPORT_MSP_PAYLOAD_SIZE   (sizeof(smartPortPayload_t) - sizeof(uint8_t))
 
-#define SMARTPORT_PAYLOAD_OFFSET offsetof(smartPortFrame_t, valueId)
-#define SMARTPORT_PAYLOAD_SIZE   (SMARTPORT_FRAME_SIZE - SMARTPORT_PAYLOAD_OFFSET - 1)
+static smartPortWriteFrameFn *smartPortWriteFrame;
 
-static smartPortFrame_t smartPortRxBuffer;
-static uint8_t smartPortRxBytes = 0;
-static bool smartPortFrameReceived = false;
 #if defined(USE_MSP_OVER_TELEMETRY)
 static bool smartPortMspReplyPending = false;
 #endif
 
-static void smartPortDataReceive(uint16_t c)
+static smartPortPayload_t *smartPortDataReceiveSerial(uint16_t c, bool *hasRequest)
```

Comment:
I think this function would be better if it returned `hasRequest` and the payload pointer was passed in as a parameter, ie
`static bool smartPortDataReceiveSerial(uint16_t c, smartPortPayload_t *payload)`